embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/intake/IntakeServiceImpl.kt

```diff
@@ -19,14 +20,23 @@ class IntakeServiceImpl(
     private val shutdownTimeoutMs: Long = 3000
 ) : IntakeService {
 
+    private var lastCacheAttempt: Future<*>? = null
+
     override fun shutdown() {
         worker.shutdownAndWait(shutdownTimeoutMs)
     }
 
     override fun take(intake: Envelope<*>, metadata: StoredTelemetryMetadata) {
-        worker.submit(metadata) {
+        val future = worker.submit(metadata) {
             processIntake(intake, metadata)
         }
+
+        // cancel any cache attempts that are already pending to avoid unnecessary I/O.
+        if (!metadata.complete) {
```

Comment:
This assumes the last cache attempt was for the same thing as the current one. I THINK we can make this assumption for now since there should be a full complete session payload if we are trying to cache a new session), but in case this were to change in the future, should we add some safeguards like checking the metadata is the same before doing so? Or just that it's an envelope of the same type?